lib/org/chromium/webidl/IDLFragmentExtractor.js

```diff
@@ -21,96 +22,190 @@ foam.CLASS({
         return retVal;
-      };
-
-      var results = lexer.parseString(this.file.contents).value;
-      if (!results) throw new Error("IDL Parse was not successful.");
-
-      var idlFragments = [];
-      var tagStack = [];            // Used for tag matching.
-      var excludeStack = [];        // Used for tracking excluded (example/note) tags.
-      var tagMatching = true;       // Set whe not inside a pre tag of interest.
-      var content = '';             // Used to group together related content.
-      for (var i = 0; i < results.length; i++) {
-        var item = results[i];
-        var isTag = lexer.Tag.isInstance(item);
-
-        // FUTURE: Handle nested pre tags.
-        // As of this writing, there has not been any IDL fragments
-        // that has been found within nested pre tags.
-        if (!tagMatching) {
-          // Ignoring all tags. Only extracting text content within pre tags.
-          if (isTag) {
-            if (item.nodeName === 'pre') {
-              tagMatching = true;
-              tagStack.pop();
-              idlFragments.push(lexer.lib.unescapeString(content));
-              content = '';
+      },
+    },
+    {
+      name: 'isExcluded',
+      documentation: 'Determines if tag has class names that excludes the tag\
+        and the descendants from IDL extraction.',
+      code: function(tag) {
+        var cls = [];
+        tag.attributes.forEach(function(attr) {
+          if (attr.name === 'class') {
+            cls = cls.concat(attr.value.split(' '));
+          }
+        });
+        return cls.includes('example') || cls.includes('note');
+      },
+    },
+    {
+      name: 'openTagAction',
+      documentation: 'Action performed when an open HTML tag is encountered.',
+      code: function(tag) {
+        var top = this.tagStack_.top;
+        var excludedTop = this.tagStack_.excludedTop;
+
+        var classes = this.extractAttr(tag, 'class');
+        var isIDL = classes && classes.includes('idl');
+
+        // HACK: Top level tags can be unmatched as well. One noticeable example
+        // is <p>. However, <div> cannot nest within <p>, so if we see an open
+        // <div> tag while <p> is on top of the stack, we pop <p>.
+        if (top && top.nodeName === 'p' && tag.nodeName === 'div') {
+          this.tagStack_.pop();
+        }
+
+        if (this.tagStack_.excludedTop === null && tag.nodeName === 'pre' && isIDL) {
+          // Found a <pre class="idl.*">.
+          // Ignore tags and only extract text now.
+          this.tagMatching_ = false;
+        }
+        this.tagStack_.push(tag);
+      },
+    },
+    {
+      name: 'extractPreContents',
+      documentation:
+        `Action performed upon encountering an open <pre> token.
+        Extraction begins at ith tag and extracts text contents until a
+        closing pre tag is encountered. If ignore is false, the extracted
+        text content is pushed into idlFragments, and the function returns
```

Comment:
The function always returns the index of the last processed element, not just if ignore is false.